### PR TITLE
maubot: init at 0.4.1

### PR DESCRIPTION
**Pull request by chayleaf**
To anyone who wants to use Maubot on NixOS, I recommend [my flake](https://github.com/chayleaf/maubot.nix) until it's merged.

###### Motivation for this change

https://github.com/maubot/maubot/

Maubot is a Matrix bot framework that has a Web UI and a plugin system. I think this PR can pave the way for a service later.

I still don't know how to handle optional dependencies though. In this case, there are optional dependencies for e2ee and Postgres. I simply haven't added them to the dependency list for now.

Previous attempt at adding it: https://github.com/NixOS/nixpkgs/pull/123760

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: maintainers/maintainer-list.nix [hunk 1]
```diff
@@ -2012,6 +2012,13 @@
     email = "me@philscotted.com";
     name = "Phil Scott";
   };
+  chayleaf = {
+    email = "chayleaf-nix@pavluk.org";
+    name = "Ivan Pavluk";
+    matrix = "chayleaf:matrix.pavluk.org";
+    github = "pavlukivan";
+    githubId = 9590981;
+  };
```
Comment (legendofmiracles) [inline maintainers/maintainer-list.nix:2021]: Should be in a different commit: `maintainers: add chayleaf`

File: pkgs/tools/networking/maubot/default.nix [hunk 2]
```diff
@@ -0,0 +1,72 @@
+{ lib
+, python3
+}:
+let python = python3.override {
```
Comment (legendofmiracles) [inline pkgs/tools/networking/maubot/default.nix:4]:
```suggestion
{ lib, python3 }:

let python = python3.override {
```

File: pkgs/tools/networking/maubot/default.nix [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    packaging
+    click
+    colorama
+    questionary
+    jinja2
+  ];
+
+  checkPhase = ''
+    $out/bin/mbc --help > /dev/null
+  '';
```
Comment (legendofmiracles) [inline pkgs/tools/networking/maubot/default.nix:59]: checkPhase is for running unit tests. For testing simple program execution, please use [package tests](https://nixos.org/manual/nixpkgs/stable/#ssec-inline-package-tests-writing)

File: pkgs/tools/networking/maubot/default.nix [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+  ];
+
+  meta = with lib; {
+    description = "A plugin-based Matrix bot system written in Python";
+    homepage = "https://github.com/maubot/maubot/";
+    license = licenses.agpl3Plus;
+    maintainers = with maintainers; [ chayleaf ];
+  };
+}
+
```
Comment (legendofmiracles) [inline pkgs/tools/networking/maubot/default.nix:72]:
```suggestion
```

File: pkgs/tools/networking/maubot/default.nix [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+    sqlalchemy = super.sqlalchemy.overridePythonAttrs (oldAttrs: rec {
+      version = "1.3.24";
+      src = oldAttrs.src.override {
+        inherit version;
+        sha256 = "06bmxzssc66cblk1hamskyv5q3xf1nh1py3vi6dka4lkpxy7gfzb";
+      };
+    });
+  };
+};
+in with python.pkgs; python.pkgs.buildPythonApplication rec {
```
Comment (legendofmiracles) [inline pkgs/tools/networking/maubot/default.nix:30]:
```suggestion
in with python.pkgs; buildPythonApplication rec {
```

File: pkgs/tools/networking/maubot/default.nix [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+  pname = "maubot";
+  version = "0.2.0";
```
Comment (legendofmiracles) [inline pkgs/tools/networking/maubot/default.nix:32]: Newest version seems to be 0.2.1

**Review by legendofmiracles — COMMENTED** [review]
Great work, thanks for your contribution.

File: pkgs/tools/networking/maubot/default.nix (hunk 2 above)
Comment (mweinelt) [reply pkgs/tools/networking/maubot/default.nix:4]: One argument per line makes for nicer diffs, so I will mark this suggestion as resolved.

Comment (legendofmiracles) [reply pkgs/tools/networking/maubot/default.nix:4]: Fair, it's something reviewers can't ever agree on :laughing:.

In general I wouldn't have made this suggestion if there had been more arguments, with only 2, things never become hard to look at in diffs imo.

It would be great if we could still get a new line after the arguments, separating it from the overlays.

File: pkgs/tools/networking/maubot/default.nix [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+
+  propagatedBuildInputs = [
+    # requirements.txt
+    mautrix
+    aiohttp
+    yarl
+    sqlalchemy
+    alembic
+    CommonMark
+    ruamel_yaml
```
Comment (mweinelt) [inline pkgs/tools/networking/maubot/default.nix:47]:
```suggestion
    ruamel-yaml
```

File: pkgs/tools/networking/maubot/default.nix (hunk 6 above)
Comment (pavlukivan) [reply pkgs/tools/networking/maubot/default.nix:32]: 0.2.1 is a docker-only release using same code as 0.2.0

**Comment by pavlukivan** [comment]
Can someone also tell me how to add optional dependencies? Or do I just assume anyone who runs the bot would want postgres and e2ee support?

File: pkgs/tools/networking/maubot/default.nix [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+    ++ optional encryptionSupport [
+      aiosqlite
+      python-olm
+      pycryptodome
+      unpaddedbase64
+    ];
+
+  passthru.tests = {
+    simple = runCommand "${pname}-tests" { } ''
+      ${maubot}/bin/mbc --help > /dev/null
```
Comment (legendofmiracles) [inline pkgs/tools/networking/maubot/default.nix:79]:
```suggestion
      ${maubot}/bin/mbc --help > $out
```

File: pkgs/tools/networking/maubot/default.nix [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+      colorama
+      questionary
+      jinja2
+    ]
+    # optional-requirements.txt
+    ++ optional postgresqlSupport [
+      psycopg2
+      asyncpg
+    ]
+    ++ optional encryptionSupport [
```
Comment (legendofmiracles) [inline pkgs/tools/networking/maubot/default.nix:70]:
```suggestion
    ++ lib.optional encryptionSupport [
```
Same goes for the other optional requirements

File: pkgs/tools/networking/maubot/default.nix [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+      unpaddedbase64
+    ];
+
+  passthru.tests = {
+    simple = runCommand "${pname}-tests" { } ''
+      ${maubot}/bin/mbc --help > /dev/null
+    '';
+  };
+
+  pythonImportsCheck = [
```
Comment (legendofmiracles) [inline pkgs/tools/networking/maubot/default.nix:83]:
```suggestion
  # python -m maubot: error: unrecognized arguments: test
  doCheck = false;

  pythonImportsCheck = [
```
Tests are failing:
```
running build_ext
usage: python -m maubot
       [-h]
       [-c <path>]
       [-b <path>]
       [-n]
python -m maubot: error: unrecognized arguments: test
```

Comment (pavlukivan) [reply pkgs/tools/networking/maubot/default.nix:83]: yeah, sorry, i can't test it just yet, will fix it tomorrow

**Comment by pavlukivan** [comment]
There's a detail I haven't accounted for. Maubot ships with both a CLI tool for managing the bot via a CLI, and a Python module which implements the actual bot functionality. There isn't a binary to run the bot, only a Python module to run via `python -m maubot`. How would I go about packaging that?

**Comment by mweinelt** [comment]
Ask the upstream to provide an entry_point. https://setuptools.pypa.io/en/latest/userguide/entry_point.html

**Comment by pavlukivan** [comment]
I used a patch

File: pkgs/tools/networking/maubot/default.nix [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "1ywvrvvq36f1bh4hl9xamj09249px0v4f0v99bgwpr7vpsxrjyhw";
+  };
+
+  # add entry point
+  patches = fetchpatch {
+    url = "https://patch-diff.githubusercontent.com/raw/maubot/maubot/pull/146.patch";
+    sha256 = "0yn5357z346qzy5v5g124mgiah1xsi9yyfq42zg028c8paiw8s8x";
+  };
```
Comment (mweinelt) [inline pkgs/tools/networking/maubot/default.nix:53]:
```suggestion
  patches = [
    (fetchpatch {
      url = "https://patch-diff.githubusercontent.com/raw/maubot/maubot/pull/146.patch";
      sha256 = "0yn5357z346qzy5v5g124mgiah1xsi9yyfq42zg028c8paiw8s8x";
    })
  ];
```

`patches` is usually list, very odd that this is supposed to work.

**Comment by mweinelt** [comment]
Please rebase.

**Comment by chayleaf** [comment]
there's a weird issue with `click` not being propagated to `lingua` that I can't quite figure out, so I'm leaving this as a draft until someone does

**Comment by jtrees** [comment]
Just wanted to chime in to say thank you for your work so far. I'm quite interested in seeing this get finished one day.

I'll poke around and see if I can help out but I'm not really familiar with the python ecosystem.

**Comment by chayleaf** [comment]
0.4.1 is out, but `mautrix` isn't updated yet in nixpkgs so I can't update this from 0.4.0

File: pkgs/tools/networking/maubot/default.nix [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+    patches = [
+      # add entry point
+      (fetchpatch {
+        url = "https://patch-diff.githubusercontent.com/raw/maubot/maubot/pull/146.patch";
+        sha256 = "0yn5357z346qzy5v5g124mgiah1xsi9yyfq42zg028c8paiw8s8x";
+      })
+      ./allow-building-plugins-from-nix-store.patch
+    ];
+
+    propagatedBuildInputs = [
```
Comment (asymmetric) [inline pkgs/tools/networking/maubot/default.nix:48]: Python noob question, but why are these `propagated` rather than just `buildInputs`?

**Comment by chayleaf** [comment]
> @chayleaf According to [the guidelines](https://nixos.org/manual/nixpkgs/stable/#contributing-guidelines), I don't think so, since it's an application,

however, if you use `python3.pkgs.maubot` and maubot is an application, you can't `import maubot`

File: pkgs/tools/networking/maubot/default.nix (hunk 12 above)
Comment (chayleaf) [reply pkgs/tools/networking/maubot/default.nix:48]: because all dependencies of maubot must also depend on those python packages so maubot itself works

File: pkgs/tools/networking/maubot/default.nix [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+      sha256 = "sha256-Ro2PPgF8818F8JewPZ3AlbfWFNNHKTZkQq+1zpm3kk4=";
+    };
+
+    patches = [
+      # add entry point
+      (fetchpatch {
+        url = "https://patch-diff.githubusercontent.com/raw/maubot/maubot/pull/146.patch";
+        sha256 = "0yn5357z346qzy5v5g124mgiah1xsi9yyfq42zg028c8paiw8s8x";
+      })
+      ./allow-building-plugins-from-nix-store.patch
```
Comment (asymmetric) [inline pkgs/tools/networking/maubot/default.nix:45]: Usually, these are in a separate dir.

```suggestion
      ./patches/allow-building-plugins-from-nix-store.patch
```

Comment (chayleaf) [reply pkgs/tools/networking/maubot/default.nix:45]:
```
$ find | grep \\.patch | wc -l
3227
$ find | grep \\.patch | grep /patches | wc -l
155
```

Comment (asymmetric) [reply pkgs/tools/networking/maubot/default.nix:45]: I stand corrected!

**Comment by asymmetric** [comment]
@chayleaf Got it. Then section 17.27.2.3.4 (which I cannot link directly) seems to apply, i.e. make it a library and use `toPythonApplication` in `all-packages.nix`.

**Comment by chayleaf** [comment]
Changed it to be in line with https://nixos.org/manual/nixpkgs/stable/#topythonapplication-function

**Comment by chayleaf** [comment]
Also I've started working on a NixOS module in the meanwhile, the WIP version is available [here](https://github.com/chayleaf/maubot.nix)

**Comment by chayleaf** [comment]
superseded by #243047
